pkg/execution/executor/executor.go

```diff
@@ -1977,13 +1996,16 @@ func (e *executor) handleGeneratorStep(ctx context.Context, i *runInstance, gen
 		MaxAttempts:           i.item.MaxAttempts,
 		Payload:               queue.PayloadEdge{Edge: nextEdge},
 	}
-	err = e.queue.Enqueue(ctx, nextItem, now, queue.EnqueueOpts{})
-	if err == redis_state.ErrQueueItemExists {
-		return nil
-	}
-	if err != nil {
-		logger.StdlibLogger(ctx).Error("error scheduling step queue item", "error", err)
-		return err
+
+	if !hasPendingSteps {
+		err = e.queue.Enqueue(ctx, nextItem, now, queue.EnqueueOpts{})
+		if err == redis_state.ErrQueueItemExists {
+			return nil
+		}
+		if err != nil {
+			logger.StdlibLogger(ctx).Error("error scheduling step queue item", "error", err)
+			return err
+		}
```

Comment:
Most enqueue calls are now wrapped in this, checking if we should continue now or whether a later execution should have that privilege.

Lifecycles are untouched and still run.